File: 0215. 数组中的第K个最大元素/Solution.java
class Solution {
    public int findKthLargest(int[] nums, int k) {

        buildHeap(nums, k);
        for(int i = k; i < nums.length; i++){
            int number = nums[i];
            if(number > nums[0]){
                nums[0] = number;
                down(0,k,nums);
            }
        }

        return nums[0];
    }

    private void buildHeap(int[] heap, int len){
        for(int i = (len - 2) / 2; i >= 0; i--){
            down(i, len, heap);
        }
    }

    

    private void down(int i, int n, int[] heap){
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        int min = i;

        if(l < n && heap[l] < heap[min])
            min = l;

        if(r < n && heap[r] < heap[min])
            min = r;

        if(min != i){
            swap(heap, min, i);
            down(min, n, heap);
        }
    }

    private void swap(int[] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: 0215. 数组中的第K个最大元素/Solution.py
class Solution:
    def partition(self, nums, l, r):
        k = random.randint(l, r)
        pivot = nums[k]
        nums[l], nums[k] = nums[k], nums[l]
        index = l
        for i in range(l + 1, r + 1):
            if nums[i] > pivot:
                index += 1
                nums[i], nums[index] = nums[index], nums[i]
        nums[l], nums[index] = nums[index], nums[l]
        return index

    def findKthLargest(self, nums: List[int], k: int) -> int:
        l, r, k = 0, len(nums) - 1, k - 1
        while 1:
            p = self.partition(nums, l, r)
            if p == k:
                return nums[p]
            elif p > k:
                r = p - 1
            else:
                l = p + 1
